File: src/tools/pcd_to_pgm/pcd_to_pgm.py
#!/usr/bin/env python3
"""
Offline PCD -> PGM converter for ROS1 map_server compatibility.

参考 ROS2 版本 /Volumes/ssd/back/zs/zs-slam_v1/src/pcd2pgm，提供：
- PCD 读取
- 可选 Z 轴截取与半径离群滤波
- 占用栅格化、膨胀
- 输出 map.pgm + map.yaml（occupied_thresh/free_thresh 兼容 map_server）
- 支持多文件目录批处理（多楼层预留）

默认参数：resolution=0.05m, inflate=1, occupied_thresh=0.65, free_thresh=0.2
"""
import argparse
import math
import os
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, List, Optional, Tuple

import numpy as np
import yaml

try:
    import open3d as o3d  # type: ignore
except ImportError:
    o3d = None


def _require_open3d():
    if o3d is None:
        raise RuntimeError("依赖 open3d，请先安装：pip install open3d")


@dataclass
class Params:
    pcd: Path
    out_dir: Path
    resolution: float = 0.05
    height_min: float = -np.inf
    height_max: float = np.inf
    inflate: int = 1
    radius: float = 0.0
    radius_min_neighbors: int = 0
    occupied_thresh: float = 0.65
    free_thresh: float = 0.2
    origin_z: float = 0.0
    transform_xyzrpy: Tuple[float, float, float, float, float, float] = (
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
    )


def load_pcd_points(pcd_path: Path) -> np.ndarray:
    _require_open3d()
    pcd = o3d.io.read_point_cloud(str(pcd_path))
    if pcd.is_empty():
        raise RuntimeError(f"PCD 空文件: {pcd_path}")
    return np.asarray(pcd.points)


def apply_transform(points: np.ndarray, xyzrpy: Tuple[float, float, float, float, float, float]) -> np.ndarray:
    tx, ty, tz, rr, rp, ry = xyzrpy
    # Rotation matrices
    cx, sx = math.cos(rr), math.sin(rr)
    cy, sy = math.cos(rp), math.sin(rp)
    cz, sz = math.cos(ry), math.sin(ry)

    Rx = np.array([[1, 0, 0], [0, cx, -sx], [0, sx, cx]])
    Ry = np.array([[cy, 0, sy], [0, 1, 0], [-sy, 0, cy]])
    Rz = np.array([[cz, -sz, 0], [sz, cz, 0], [0, 0, 1]])
    R = Rz @ Ry @ Rx
    t = np.array([tx, ty, tz])
    return (R @ points.T).T + t


def passthrough_filter(points: np.ndarray, z_min: float, z_max: float) -> np.ndarray:
    mask = (points[:, 2] >= z_min) & (points[:, 2] <= z_max)
    return points[mask]


def radius_outlier_filter(points: np.ndarray, radius: float, min_neighbors: int) -> np.ndarray:
    if radius <= 0 or min_neighbors <= 0:
        return points
    _require_open3d()
    pcd = o3d.geometry.PointCloud(o3d.utility.Vector3dVector(points))
    _, ind = pcd.remove_radius_outlier(nb_points=min_neighbors, radius=radius)
    return points[ind]


def compute_bounds_xy(points: np.ndarray) -> Tuple[float, float, float, float]:
    x_min, y_min = np.min(points[:, 0]), np.min(points[:, 1])
    x_max, y_max = np.max(points[:, 0]), np.max(points[:, 1])
    return x_min, x_max, y_min, y_max


def points_to_grid(points: np.ndarray, resolution: float, inflate: int) -> Tuple[np.ndarray, Tuple[float, float]]:
    x_min, x_max, y_min, y_max = compute_bounds_xy(points)
    width = int(math.ceil((x_max - x_min) / resolution)) + 1
    height = int(math.ceil((y_max - y_min) / resolution)) + 1

    grid = np.zeros((height, width), dtype=np.uint8)  # 0=free, 100=occupied, 255=unknown
    # unknown 初始化为 255 符合 map_server 读取，但最终 PGM 会重新映射
    grid.fill(255)

    xi = np.floor((points[:, 0] - x_min) / resolution).astype(np.int32)
    yi = np.floor((points[:, 1] - y_min) / resolution).astype(np.int32)

    grid[yi, xi] = 100

    if inflate > 0:
        grid = inflate_grid(grid, inflate)

    origin = (x_min, y_min)
    return grid, origin


def inflate_grid(grid: np.ndarray, radius: int) -> np.ndarray:
    padded = np.pad(grid, radius, constant_values=255)
    h, w = grid.shape
    inflated = grid.copy()
    for dy in range(-radius, radius + 1):
        for dx in range(-radius, radius + 1):
            if dx == 0 and dy == 0:
                continue
            sub = padded[radius + dy : radius + dy + h, radius + dx : radius + dx + w]
            inflated = np.minimum(inflated, sub)
    return inflated


def grid_to_pgm(grid: np.ndarray, out_path: Path) -> None:
    # map_server: occupied=0, free=254, unknown=205
    pgm_grid = np.full_like(grid, 205, dtype=np.uint8)
    pgm_grid[grid == 100] = 0
    pgm_grid[grid == 255] = 205
    pgm_grid[(grid != 100) & (grid != 255)] = 254

    h, w = pgm_grid.shape
    header = f"P5\n{w} {h}\n255\n"
    with open(out_path, "wb") as f:
        f.write(header.encode("ascii"))
        f.write(np.flipud(pgm_grid).tobytes())  # flip Y so map origin at bottom-left


def write_yaml(yaml_path: Path, pgm_path: Path, resolution: float, origin: Tuple[float, float], origin_z: float, occupied_thresh: float, free_thresh: float) -> None:
    data = {
        "image": pgm_path.name,
        "resolution": float(resolution),
        "origin": [float(origin[0]), float(origin[1]), float(origin_z)],
        "negate": 0,
        "occupied_thresh": float(occupied_thresh),
        "free_thresh": float(free_thresh),
    }
    with open(yaml_path, "w", encoding="utf-8") as f:
        yaml.safe_dump(data, f, default_flow_style=False)


def process_single(params: Params) -> Path:
    points = load_pcd_points(params.pcd)
    points = apply_transform(points, params.transform_xyzrpy)
    points = passthrough_filter(points, params.height_min, params.height_max)
    points = radius_outlier_filter(points, params.radius, params.radius_min_neighbors)

    if points.shape[0] == 0:
        raise RuntimeError(f"过滤后点云为空: {params.pcd}")

    grid, origin = points_to_grid(points, params.resolution, params.inflate)
    params.out_dir.mkdir(parents=True, exist_ok=True)
    pgm_path = params.out_dir / "map.pgm"
    yaml_path = params.out_dir / "map.yaml"
    grid_to_pgm(grid, pgm_path)
    write_yaml(yaml_path, pgm_path, params.resolution, origin, params.origin_z, params.occupied_thresh, params.free_thresh)
    return yaml_path


def process_directory(pcd_dir: Path, out_dir: Path, base_params: Params) -> List[Path]:
    outputs: List[Path] = []
    pcd_files = sorted(p for p in pcd_dir.iterdir() if p.suffix.lower() == ".pcd")
    if not pcd_files:
        raise RuntimeError(f"目录下没有 PCD 文件: {pcd_dir}")
    index = []
    for idx, pcd_path in enumerate(pcd_files):
        floor_id = idx
        floor_dir = out_dir / f"floor_{floor_id}"
        yaml_path = process_single(
            Params(
                pcd=pcd_path,
                out_dir=floor_dir,
                resolution=base_params.resolution,
                height_min=base_params.height_min,
                height_max=base_params.height_max,
                inflate=base_params.inflate,
                radius=base_params.radius,
                radius_min_neighbors=base_params.radius_min_neighbors,
                occupied_thresh=base_params.occupied_thresh,
                free_thresh=base_params.free_thresh,
                origin_z=base_params.origin_z,
                transform_xyzrpy=base_params.transform_xyzrpy,
            )
        )
        outputs.append(yaml_path)
        index.append(
            {
                "floor": floor_id,
                "pcd": str(pcd_path.name),
                "map": str(yaml_path.relative_to(out_dir)),
            }
        )
    index_yaml = out_dir / "index.yaml"
    with open(index_yaml, "w", encoding="utf-8") as f:
        yaml.safe_dump(index, f, default_flow_style=False)
    return outputs


def parse_args(argv: Optional[Iterable[str]] = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Offline PCD -> PGM converter (ROS1 map_server compatible)")
    parser.add_argument("--pcd", type=Path, required=True, help="PCD 文件路径或目录（目录则批处理，多楼层输出）")
    parser.add_argument("--out", type=Path, required=True, help="输出目录")
    parser.add_argument("--resolution", type=float, default=0.05, help="栅格分辨率，米")
    parser.add_argument("--height_min", type=float, default=-np.inf, help="Z 下界")
    parser.add_argument("--height_max", type=float, default=np.inf, help="Z 上界")
    parser.add_argument("--inflate", type=int, default=1, help="膨胀半径，单位：cell")
    parser.add_argument("--radius", type=float, default=0.0, help="半径离群滤波的半径，<=0 则跳过")
    parser.add_argument("--radius_min_neighbors", type=int, default=0, help="半径滤波最少邻居数，<=0 跳过")
    parser.add_argument("--occupied_thresh", type=float, default=0.65, help="YAML occupied_thresh")
    parser.add_argument("--free_thresh", type=float, default=0.2, help="YAML free_thresh")
    parser.add_argument("--origin_z", type=float, default=0.0, help="YAML 原点 z")
    parser.add_argument(
        "--transform",
        type=float,
        nargs=6,
        default=(0, 0, 0, 0, 0, 0),
        metavar=("x", "y", "z", "roll", "pitch", "yaw"),
        help="先旋转后平移（米/弧度），与 ROS2 组件 odom_to_lidar_odom 一致",
    )
    return parser.parse_args(argv)


def main(argv: Optional[Iterable[str]] = None) -> int:
    args = parse_args(argv)
    pcd_path: Path = args.pcd
    out_dir: Path = args.out

    base_params = Params(
        pcd=pcd_path,
        out_dir=out_dir,
        resolution=args.resolution,
        height_min=args.height_min,
        height_max=args.height_max,
        inflate=args.inflate,
        radius=args.radius,
        radius_min_neighbors=args.radius_min_neighbors,
        occupied_thresh=args.occupied_thresh,
        free_thresh=args.free_thresh,
        origin_z=args.origin_z,
        transform_xyzrpy=tuple(args.transform),
    )
    try:
        if pcd_path.is_dir():
            process_directory(pcd_path, out_dir, base_params)
        else:
            process_single(base_params)
    except Exception as e:  # pylint: disable=broad-except
        print(f"[pcd_to_pgm] 失败: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/tools/pcd_to_pgm/pcd_to_pgm.cpp
#include <Eigen/Geometry>
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <cmath>
#include <string>
#include <tuple>
#include <vector>

#include <pcl/common/transforms.h>
#include <pcl/filters/passthrough.h>
#include <pcl/filters/radius_outlier_removal.h>
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>

namespace fs = std::filesystem;

struct Params
{
  fs::path pcd_path;
  fs::path out_dir;
  double resolution{0.05};
  double z_min{-0.5};
  double z_max{0.5};
  bool invert_z{false};  // same meaning as setNegative in ROS2 node
  double radius{0.1};
  int min_neighbors{10};
  int inflate_radius{0};
  int min_points_per_cell{1};  // 聚类阈值，过滤稀疏轨迹点
  int open_radius{0};          // 形态学 opening 半径，去除细线条
  double origin_z{0.0};
  double occupied_thresh{0.65};
  double free_thresh{0.2};
  double transform[6]{0.0, 0.0, 0.0, 0.0, 0.0, 0.0};  // x y z roll pitch yaw
};

struct Bounds
{
  double x_min{std::numeric_limits<double>::max()};
  double x_max{std::numeric_limits<double>::lowest()};
  double y_min{std::numeric_limits<double>::max()};
  double y_max{std::numeric_limits<double>::lowest()};
};

struct Grid
{
  std::vector<uint8_t> data;  // 0=free, 100=occupied, 255=unknown
  int width{0};
  int height{0};
  double origin_x{0.0};
  double origin_y{0.0};
};

void printUsage()
{
  std::cout << "PCD -> PGM offline converter (C++), ROS1/ROS2 map_server friendly\n"
            << "Usage:\n"
            << "  pcd_to_pgm_cpp --pcd <file_or_dir> --out <dir> [options]\n\n"
            << "Options:\n"
            << "  --resolution <m>          Grid resolution (default 0.05)\n"
            << "  --z-min <val>             PassThrough lower bound (default -inf)\n"
            << "  --z-max <val>             PassThrough upper bound (default +inf)\n"
            << "  --invert-z                Keep points OUTSIDE [z-min, z-max] (default: keep inside)\n"
            << "  --radius <val>            Radius outlier removal radius (<=0 to disable)\n"
            << "  --min-neighbors <n>       Min neighbors in radius filter (<=0 to disable)\n"
            << "  --inflate <cells>         Obstacle inflation radius in cells (default 1)\n"
            << "  --min-points-per-cell <n> Min points needed to mark a cell occupied (default 1)\n"
            << "  --open-radius <cells>     Morphological opening radius to remove thin trails (default 0)\n"
            << "  --origin-z <val>          YAML origin z (default 0)\n"
            << "  --occupied-thresh <val>   YAML occupied_thresh (default 0.65)\n"
            << "  --free-thresh <val>       YAML free_thresh (default 0.2)\n"
            << "  --transform x y z r p y   Apply RPY then translate, same as odom_to_lidar_odom\n"
            << "  --help                    Show this message\n";
}

bool parseArguments(int argc, char ** argv, Params & params)
{
  if (argc < 2) {
    printUsage();
    return false;
  }

  auto requireValue = [&](int & idx) -> std::string {
    if (idx + 1 >= argc) {
      throw std::runtime_error("Option missing value: " + std::string(argv[idx]));
    }
    return std::string(argv[++idx]);
  };

  try {
    for (int i = 1; i < argc; ++i) {
      std::string arg = argv[i];
      if (arg == "--pcd") {
        params.pcd_path = requireValue(i);
      } else if (arg == "--out") {
        params.out_dir = requireValue(i);
      } else if (arg == "--resolution") {
        params.resolution = std::stod(requireValue(i));
      } else if (arg == "--z-min") {
        params.z_min = std::stod(requireValue(i));
      } else if (arg == "--z-max") {
        params.z_max = std::stod(requireValue(i));
      } else if (arg == "--invert-z") {
        params.invert_z = true;
      } else if (arg == "--radius") {
        params.radius = std::stod(requireValue(i));
      } else if (arg == "--min-neighbors") {
        params.min_neighbors = std::stoi(requireValue(i));
      } else if (arg == "--inflate") {
        params.inflate_radius = std::stoi(requireValue(i));
      } else if (arg == "--min-points-per-cell") {
        params.min_points_per_cell = std::stoi(requireValue(i));
      } else if (arg == "--open-radius") {
        params.open_radius = std::stoi(requireValue(i));
      } else if (arg == "--origin-z") {
        params.origin_z = std::stod(requireValue(i));
      } else if (arg == "--occupied-thresh") {
        params.occupied_thresh = std::stod(requireValue(i));
      } else if (arg == "--free-thresh") {
        params.free_thresh = std::stod(requireValue(i));
      } else if (arg == "--transform") {
        for (int k = 0; k < 6; ++k) {
          params.transform[k] = std::stod(requireValue(i));
        }
      } else if (arg == "--help" || arg == "-h") {
        printUsage();
        return false;
      } else {
        throw std::runtime_error("Unknown option: " + arg);
      }
    }
  } catch (const std::exception & e) {
    std::cerr << "Argument error: " << e.what() << "\n";
    printUsage();
    return false;
  }

  if (params.pcd_path.empty() || params.out_dir.empty()) {
    std::cerr << "Missing required --pcd and/or --out\n";
    printUsage();
    return false;
  }

  return true;
}

pcl::PointCloud<pcl::PointXYZ>::Ptr loadPcd(const fs::path & pcd_path)
{
  auto cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  if (pcl::io::loadPCDFile<pcl::PointXYZ>(pcd_path.string(), *cloud) != 0) {
    throw std::runtime_error("Failed to load PCD: " + pcd_path.string());
  }
  if (cloud->empty()) {
    throw std::runtime_error("PCD is empty: " + pcd_path.string());
  }
  return cloud;
}

void applyTransform(pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud, const double transform[6])
{
  Eigen::Affine3f tf = Eigen::Affine3f::Identity();
  tf.translation() << static_cast<float>(transform[0]), static_cast<float>(transform[1]),
    static_cast<float>(transform[2]);
  tf.rotate(Eigen::AngleAxisf(transform[3], Eigen::Vector3f::UnitX()));
  tf.rotate(Eigen::AngleAxisf(transform[4], Eigen::Vector3f::UnitY()));
  tf.rotate(Eigen::AngleAxisf(transform[5], Eigen::Vector3f::UnitZ()));

  // Keep consistent with ROS2 node (odom_to_lidar_odom): apply inverse
  pcl::transformPointCloud(*cloud, *cloud, tf.inverse());
}

pcl::PointCloud<pcl::PointXYZ>::Ptr passThroughFilter(
  const pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud, double z_min, double z_max, bool invert)
{
  auto filtered = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  pcl::PassThrough<pcl::PointXYZ> pass;
  pass.setInputCloud(cloud);
  pass.setFilterFieldName("z");
  pass.setFilterLimits(z_min, z_max);
  pass.setNegative(invert);
  pass.filter(*filtered);
  return filtered;
}

pcl::PointCloud<pcl::PointXYZ>::Ptr radiusOutlierFilter(
  const pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud, double radius, int min_neighbors)
{
  if (radius <= 0.0 || min_neighbors <= 0) {
    return cloud;
  }
  auto filtered = std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  pcl::RadiusOutlierRemoval<pcl::PointXYZ> radius_filter;
  radius_filter.setInputCloud(cloud);
  radius_filter.setRadiusSearch(radius);
  radius_filter.setMinNeighborsInRadius(min_neighbors);
  radius_filter.filter(*filtered);
  return filtered;
}

Bounds computeBounds(const pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud)
{
  Bounds b;
  for (const auto & pt : cloud->points) {
    b.x_min = std::min(b.x_min, static_cast<double>(pt.x));
    b.x_max = std::max(b.x_max, static_cast<double>(pt.x));
    b.y_min = std::min(b.y_min, static_cast<double>(pt.y));
    b.y_max = std::max(b.y_max, static_cast<double>(pt.y));
  }
  return b;
}

Grid buildGrid(
  const pcl::PointCloud<pcl::PointXYZ>::Ptr & cloud, double resolution, int inflate_radius,
  int min_points_per_cell, int open_radius)
{
  if (cloud->empty()) {
    throw std::runtime_error("Cloud empty after filtering");
  }

  Bounds b = computeBounds(cloud);
  const double epsilon = 1e-6;
  b.x_min -= epsilon;
  b.y_min -= epsilon;

  int width = static_cast<int>(std::ceil((b.x_max - b.x_min) / resolution));
  int height = static_cast<int>(std::ceil((b.y_max - b.y_min) / resolution));
  std::vector<uint8_t> grid(width * height, 0);  // free by default
  std::vector<int> counts(width * height, 0);

  for (const auto & pt : cloud->points) {
    int ix = static_cast<int>(std::floor((pt.x - b.x_min) / resolution));
    int iy = static_cast<int>(std::floor((pt.y - b.y_min) / resolution));
    if (ix >= 0 && ix < width && iy >= 0 && iy < height) {
      counts[ix + iy * width] += 1;
    }
  }

  for (int idx = 0; idx < width * height; ++idx) {
    if (counts[idx] >= min_points_per_cell) {
      grid[idx] = 100;
    }
  }

  auto erode = [&](const std::vector<uint8_t> & src, std::vector<uint8_t> & dst, int radius) {
    dst.assign(src.size(), 0);
    for (int y = 0; y < height; ++y) {
      for (int x = 0; x < width; ++x) {
        bool keep = true;
        for (int dy = -radius; dy <= radius && keep; ++dy) {
          for (int dx = -radius; dx <= radius; ++dx) {
            int nx = x + dx;
            int ny = y + dy;
            if (nx < 0 || nx >= width || ny < 0 || ny >= height ||
                src[nx + ny * width] != 100) {
              keep = false;
              break;
            }
          }
        }
        if (keep) {
          dst[x + y * width] = 100;
        }
      }
    }
  };

  auto dilate = [&](const std::vector<uint8_t> & src, std::vector<uint8_t> & dst, int radius) {
    dst.assign(src.size(), 0);
    for (int y = 0; y < height; ++y) {
      for (int x = 0; x < width; ++x) {
        bool hit = false;
        for (int dy = -radius; dy <= radius && !hit; ++dy) {
          for (int dx = -radius; dx <= radius; ++dx) {
            int nx = x + dx;
            int ny = y + dy;
            if (nx >= 0 && nx < width && ny >= 0 && ny < height &&
                src[nx + ny * width] == 100) {
              hit = true;
              break;
            }
          }
        }
        if (hit) {
          dst[x + y * width] = 100;
        }
      }
    }
  };

  if (open_radius > 0) {
    std::vector<uint8_t> tmp;
    erode(grid, tmp, open_radius);
    dilate(tmp, grid, open_radius);
  }

  if (inflate_radius > 0) {
    std::vector<uint8_t> inflated(grid);
    for (int y = 0; y < height; ++y) {
      for (int x = 0; x < width; ++x) {
        if (grid[x + y * width] != 100) {
          continue;
        }
        for (int dy = -inflate_radius; dy <= inflate_radius; ++dy) {
          for (int dx = -inflate_radius; dx <= inflate_radius; ++dx) {
            int nx = x + dx;
            int ny = y + dy;
            if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
              inflated[nx + ny * width] = 100;
            }
          }
        }
      }
    }
    grid.swap(inflated);
  }

  return Grid{std::move(grid), width, height, b.x_min, b.y_min};
}

void writePgm(const Grid & grid, const fs::path & pgm_path)
{
  std::ofstream ofs(pgm_path, std::ios::binary);
  if (!ofs) {
    throw std::runtime_error("Failed to open PGM for write: " + pgm_path.string());
  }
  ofs << "P5\n" << grid.width << " " << grid.height << "\n255\n";

  // Flip Y to match map_server origin at bottom-left
  for (int y = grid.height - 1; y >= 0; --y) {
    for (int x = 0; x < grid.width; ++x) {
      uint8_t cell = grid.data[x + y * grid.width];
      uint8_t pixel = 254;  // free
      if (cell == 100) {
        pixel = 0;  // occupied
      } else if (cell == 255) {
        pixel = 205;  // unknown
      }
      ofs.put(static_cast<char>(pixel));
    }
  }
}

void writeYaml(
  const fs::path & yaml_path, const fs::path & pgm_path, const Grid & grid, const Params & params)
{
  std::ofstream ofs(yaml_path);
  if (!ofs) {
    throw std::runtime_error("Failed to open YAML for write: " + yaml_path.string());
  }
  ofs << std::fixed << std::setprecision(6);
  ofs << "image: " << pgm_path.filename().string() << "\n";
  ofs << "resolution: " << params.resolution << "\n";
  ofs << "origin: [" << grid.origin_x << ", " << grid.origin_y << ", " << params.origin_z << "]\n";
  ofs << "negate: 0\n";
  ofs << "occupied_thresh: " << params.occupied_thresh << "\n";
  ofs << "free_thresh: " << params.free_thresh << "\n";
}

struct ConversionResult
{
  fs::path yaml_path;
  fs::path pgm_path;
};

ConversionResult convertOne(const fs::path & pcd_path, const fs::path & out_dir, const Params & p)
{
  auto cloud = loadPcd(pcd_path);
  applyTransform(cloud, p.transform);

  cloud = passThroughFilter(cloud, p.z_min, p.z_max, p.invert_z);
  cloud = radiusOutlierFilter(cloud, p.radius, p.min_neighbors);

  if (cloud->empty()) {
    throw std::runtime_error("Cloud empty after filtering: " + pcd_path.string());
  }

  Grid grid =
    buildGrid(cloud, p.resolution, p.inflate_radius, p.min_points_per_cell, p.open_radius);
  fs::create_directories(out_dir);
  fs::path pgm_path = out_dir / "map.pgm";
  fs::path yaml_path = out_dir / "map.yaml";
  writePgm(grid, pgm_path);
  writeYaml(yaml_path, pgm_path, grid, p);
  return {yaml_path, pgm_path};
}

void writeIndexYaml(
  const fs::path & index_path, const std::vector<std::tuple<int, fs::path, fs::path>> & entries)
{
  std::ofstream ofs(index_path);
  if (!ofs) {
    throw std::runtime_error("Failed to open index.yaml for write: " + index_path.string());
  }
  for (const auto & [floor_id, pcd_name, map_yaml] : entries) {
    ofs << "- floor: " << floor_id << "\n";
    ofs << "  pcd: " << pcd_name.filename().string() << "\n";
    ofs << "  map: " << map_yaml.string() << "\n";
  }
}

int main(int argc, char ** argv)
{
  Params params;
  if (!parseArguments(argc, argv, params)) {
    return 1;
  }

  try {
    if (fs::is_directory(params.pcd_path)) {
      std::vector<fs::path> pcd_files;
      for (const auto & entry : fs::directory_iterator(params.pcd_path)) {
        if (entry.is_regular_file() && entry.path().extension() == ".pcd") {
          pcd_files.push_back(entry.path());
        }
      }
      std::sort(pcd_files.begin(), pcd_files.end());
      if (pcd_files.empty()) {
        throw std::runtime_error("No .pcd files found in directory");
      }

      std::vector<std::tuple<int, fs::path, fs::path>> index_entries;
      for (size_t i = 0; i < pcd_files.size(); ++i) {
        fs::path floor_dir = params.out_dir / ("floor_" + std::to_string(i));
        auto result = convertOne(pcd_files[i], floor_dir, params);
        index_entries.emplace_back(
          static_cast<int>(i), pcd_files[i].filename(),
          fs::relative(result.yaml_path, params.out_dir));
        std::cout << "[pcd_to_pgm_cpp] Converted " << pcd_files[i] << " -> " << result.yaml_path
                  << "\n";
      }
      writeIndexYaml(params.out_dir / "index.yaml", index_entries);
      std::cout << "[pcd_to_pgm_cpp] Wrote index.yaml with " << index_entries.size()
                << " entries\n";
    } else {
      auto result = convertOne(params.pcd_path, params.out_dir, params);
      std::cout << "[pcd_to_pgm_cpp] Converted -> " << result.yaml_path << "\n";
    }
  } catch (const std::exception & e) {
    std::cerr << "[pcd_to_pgm_cpp] Error: " << e.what() << "\n";
    return 1;
  }

  return 0;
}
